utils/s3/credentials_providers/sts_assume_role_credentials_provider.cc

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "sts_assume_role_credentials_provider.hh"
+
+#include "utils/UUID.hh"
+#include "utils/http.hh"
+#include "utils/s3/client.hh"
+#include <rapidxml.h>
+#include <seastar/core/coroutine.hh>
+#include <seastar/http/request.hh>
+#include <seastar/util/short_streams.hh>
+
+namespace aws {
+
+static logging::logger sts_logger("sts");
+
+sts_assume_role_credentials_provider::sts_assume_role_credentials_provider(const std::string& _host, unsigned _port, bool _is_secured)
+    : sts_host(_host), port(_port), is_secured(_is_secured) {
+}
+
+sts_assume_role_credentials_provider::sts_assume_role_credentials_provider(const std::string& _region, const std::string& _role_arn)
+    : sts_host(seastar::format("sts.{}.amazonaws.com", _region)), role_arn(_role_arn) {
+}
+
+future<s3::aws_credentials> sts_assume_role_credentials_provider::get_aws_credentials() {
+    co_await reload();
+    co_return creds;
+}
+
+bool sts_assume_role_credentials_provider::is_time_to_refresh() const {
+    return std::chrono::system_clock::now() >= creds.expires_at;
```

Comment:
Can be lowres_clock. Expiration times are hours and there's no need in using high-precision timepoints